indico/modules/events/contributions/controllers/display.py

```diff
@@ -97,9 +98,17 @@ def _check_access(self):
             raise Forbidden
 
     def _process(self):
-        contributions = get_contributions_with_user_as_submitter(self.event, session.user)
+        contribs = get_contributions_for_user(self.event, session.user)
+        categorized = defaultdict(list)
+        for contrib in contribs:
+            if any(link.person.user == session.user for link in contrib.primary_authors):
+                categorized['primary'].append(contrib)
+            if any(link.person.user == session.user for link in contrib.secondary_authors):
+                categorized['secondary'].append(contrib)
+            if any(link.person.user == session.user for link in contrib.speakers):
+                categorized['speaker'].append(contrib)
```

Comment:
You could iterate directly over `contrib.person_links` and then just check the roles (`author_type` & `is_speaker`) of that object.